aips/aip-x.md

```diff
@@ -0,0 +1,237 @@
+
+Since the signer is not storable, we should have the freedom to change the internal representation of a signer. The new signer will roughly look like the following in the VM:
+
+```rust
+enum Signer {
+    Master(AccountAddress),
+    Permissioned {
+        master_addr: address,
+        permission_addr: address,
+    }
+}
+```
+
+Note that we did change the serialization rules for signer which would be a non-backward compatible change. This should be fine in general as signer should not be storable. The only problem is we exposes `size_of_val`  API in our framework. Since we changed the serialization, this would break the existing semantics for this API. To mitigate this, we will manually make sure that  `size_of_val`  would calculate the size of signer to be a single account address instead of an address plus a variant.
+
+### Managing permissions
+
+The management API would look like the following:
+
+```jsx
+module aptos_framework::permissioned_signer {
+    public fun authorize<PermKey: copy + drop + store>(
+        master: &signer,
+        permissioned: &signer,
+        capacity: u256,
+        perm: PermKey
+    );
+
+    public fun check_permission_exists<PermKey: copy + drop + store>(s: &signer, perm: PermKey)
+
+    public fun check_permission_capacity_above<PermKey: copy + drop + store>(
+        s: &signer, threshold: u256, perm: PermKey
+    )
+
+    public fun check_permission_consume<PermKey: copy + drop + store>(
+        s: &signer, weight: u256, perm: PermKey
+    )
+}
+```
+
+The master signer can use `authorize`  to grant permissions to a permissioned signer. The smart contracts would use the `check_` APIs to make sure if a permissioned signer is provided, it needs to have the right permission granted by the master signer to perform the corresponding operation.
+
+In the current design, the permission storage is a `CopyableAny -> u256`  mapping stored in the global storage. Smart contract can define the key themselves to create proper domain separation and act as witness pattern.
+
+### Revoking Permission
+
+The master signer should be able to revoke the permission granted to a permissioned signer at any time. The API would look like the following:
+
+```jsx
+module aptos_framework::permissioned_signer {
+    public fun revoke_permission_handle(master: &signer, permission_address: address);
+}
+```
+
+The `signer_from_permissioned` would abort on revoked signer.
+
+### Example: Fungible Asset Transfer Limit
+```jsx
+module aptos_framework::fungible_asset {
+    struct WithdrawPermission has store {
+        token_addr: address,
+    }
+
+    public fun withdraw<T: key>(owner: &signer, store: Object<T>, amount: u64): FungibleAsset {
+        assert!(
+            permissioned_signer::check_permission(
+                owner,
+                amount as u256,
+                WithdrawPermission {
+                    metadata_address: object::object_address(&borrow_store_resource(&store).metadata)
+                }
```

Comment:
Maybe make `WithdrawPermission` an enum and place amount in the enum?
Do the casts to u256 cost extra?

Reply:
The amount has to be in the value storage as this needs to be updated upon withdraw. We do plan to make a couple of items there to be enum tho once the enum is stablized.